sway-lib-std/src/ufp64.sw

```diff
@@ -0,0 +1,247 @@
+    pub fn divide(self, divisor: Self) -> Self {
+    let zero = ~UFP64::zero();
+        assert(divisor != zero);
+
+        let denominator = ~U128::from(0, ~Self::denominator());
+        let self_u128 = ~U128::from(0, self.value);
+        let divisor_u128 = ~U128::from(0, divisor.value);
+
+        let res_u128 = self_u128 * denominator / divisor_u128;
+        if res_u128.upper != 0 {
+            // panic on overflow
+            revert(0);
+        }
+        UFP64 {
+            value: res_u128.lower
+        }
+    }
+}
+
+
+
+impl UFP64 {
+
+    pub fn from_uint(uint: u64) -> UFP64 {
+        UFP64 {
+            value: ~Self::denominator() * uint,
+        }
+    }
+}
+
+impl UFP64 {
+
+    pub fn recip(number: UFP64) -> Self {
+        let one = ~UFP64::from_uint(1);
+        
+        let res = one / number;
+        res
+    }
+}
+
+impl UFP64 {
+
+    pub fn trunc(self) -> Self {
+        UFP64 {
+            value: (self.value >> 32) << 32
+        }
+    }
+}
+
+impl UFP64 {
+
+    pub fn floor(self) -> Self {
+        return self.trunc();
+    }
+
+    pub fn fract(self) -> Self {
+        UFP64 {
+            value: (self.value << 32) >> 32
+        }
+    }
+}
+
+impl UFP64 {
+
+    pub fn ceil(self) -> Self {
+        if self.fract().value != 0 {
+            let res = self.trunc() + ~UFP64::from_uint(1);
+            return res;
+        }
+        return self;
+    }
+}
+
+impl UFP64 {
+
+    pub fn round(self) -> Self {
+        let floor = self.floor();
+        let ceil = self.ceil();
+        let diff_self_floor = self - floor;
+        let diff_ceil_self = ceil - self;
+        if diff_self_floor < diff_ceil_self {
+            return floor;
+        } else {
+            return ceil;
+        }
+    }  
+}
+
+impl Root for UFP64 {
+    fn sqrt(self) -> Self {
+        let nominator_root = self.value.sqrt();
+        let nominator = nominator_root << 16;
+        UFP64 {
+            value: nominator
+        }
+    }
+}
+
+impl Exponentiate for UFP64 {
+    pub fn pow(self, exponent: Self) -> Self {
+        let denominator_power = 32;
+        let exponent_int = exponent.value >> denominator_power;
+        let nominator_pow = ~U128::from(0, self.value).pow(~U128::from(0, exponent_int));
+        let nominator = nominator_pow >> denominator_power*(exponent_int - 1);
+        if nominator.upper != 0 {
+            // panic on overflow
+            revert(0);
+        }
+        UFP64 {
+            value: nominator.lower
+        }
+    }
+}
+
+// TODO: uncomment and change accordingly, when signed integers will be added
+// impl Logarithm for UFP64 {
+//     fn log(self, base: Self) -> Self {
+//         let nominator_log = self.value.log(base);
+//         let res = (nominator_log - ~u64::from(0, 64 * 2.log(base))) * ~u64::from(1, 0);
+//         UFP64 {
+//             value: res
+//         }
+//     }
+// }
+
+impl Exponent for UFP64 {
+    pub fn exp(exponent: Self) -> Self {
```

Comment:
This looks disgustingly evil. Can this be cleared up somehow?

Reply:
Changed it to exact coeficients. Hope that solved the issue

Reply:
I was specifically referring to all of these variables being declared and that LONG calculation below. I do prefer the name "from" however those magic numbers are hard to reason about now. Can you add some comments to indicate what they represent?

Reply:
I hoped earlier declaration explained them. These coefficients are just reciprocals to factorials of the indeces. E.g. p2 == 1 / 2! and p6 == 1 / 6!
Calculating them beforehand is a better idea, so unfortunately we need to have magic numbers. I'll add these comments to the code.
The long calculation is a common approximation  of exp (e.g. c++ libs use it) using Taylor series up to 7th power. I don't thing there is a faster way to calculate this.

Reply:
Added the comments to the code